string.go

```diff
@@ -4,29 +4,52 @@
 
 package validation
 
-import "errors"
-
 type stringValidator func(string) bool
 
 // StringRule is a rule that checks a string variable using a specified stringValidator.
 type StringRule struct {
 	validate stringValidator
-	message  string
+	err      Error
 }
 
 // NewStringRule creates a new validation rule using a function that takes a string value and returns a bool.
 // The rule returned will use the function to check if a given string or byte slice is valid or not.
 // An empty value is considered to be valid. Please use the Required rule to make sure a value is not empty.
+//
+// Deprecated: This function has been deprecated in favor of NewStringValidator. it exists for historical
+// compatibility and should not be used.
 func NewStringRule(validator stringValidator, message string) StringRule {
 	return StringRule{
 		validate: validator,
-		message:  message,
+		err:      NewError("", message),
+	}
+}
+
+// NewStringValidator creates a new validation rule using a function that takes a string value and returns a bool.
+// The rule returned will use the function to check if a given string or byte slice is valid or not.
+// An empty value is considered to be valid. Please use the Required rule to make sure a value is not empty.
+func NewStringValidator(validator stringValidator, code, message string) StringRule {
+	return StringRule{
+		validate: validator,
+		err:      NewError(code, message),
 	}
 }
 
 // Error sets the error message for the rule.
 func (v StringRule) Error(message string) StringRule {
-	v.message = message
+	v.err = v.err.SetMessage(message)
+	return v
+}
+
```

Comment:
How about adding `ErrorStruct() *Error` and drop the following two methods?

Reply:
I think we should keep the following two methods and also add this function.

Reply:
I think we should keep those two functions and add this function also.

Reply:
Why keeping the two functions? If we add the new one, developers can first get `Error` and then call the methods of `Error` to set Code, etc.

Reply:
because if we keep those functions, developers need to one call for setting custom error message, by defining this function need to two call, if we define two functions one for setting message, another for getting error, developers can simply set message just like old version, and also can get error if want to set params or any other thing.

I think we should add `ErrorStruct() *Error` to all built-in validation rules, so developers can set custom error codes on each validation.

Reply:
I thought it more about the overall translation support. Below I'm using `ThresholdRule` as an example to explain the new changes needed since `ThresholdRule` is one of the most complex rules for the translation support:
```go
var (
    // predefined errors that can be customized globally to affect all instances of the same type of the rule
    // used in both i18n and non-i18n scenarios
    ErrThresholdLessThanRequired = NewError("validation_threshold_less_than_required", "must be less than {{.threshold}}")
    ErrThresholdNoGreaterThanRequired = NewError("validation_threshold_no_greater_than_required", "must be no greater than {{.threshold}}")
    ErrThresholdGreaterThanRequired = NewError("validation_threshold_greater_than_required", "must be greater than {{.threshold}}")
    ErrThresholdNoLessThanRequired = NewError("validation_threshold_no_less_than_required", "must be no less than {{.threshold}}")
)

// used in non-i18n scenarios to customize an error message for a single instance of the rule
func (r ThresholdRule) Error(message string) ThresholdRule {...}

// used in i18n scenarios to customize the error for a single instance of the rule
func (r ThresholdRule) ErrorObject(err error) ThresholdRule {...}
```

`StringRule` is a bit special because it is a base class. Here's the change I'm thinking about:
```go
// used in non-i18n scenarios
func NewStringRule(validator stringValidator, message string) StringRule {...}

// used in i18n scenarios
func NewStringRuleWithError(validator stringValidator, err error) StringRule {...}

// This error declaration allows global customization of the email rule error
// String.ErrorObject() allows error customization of a single rule instance
ErrEmail = NewError("validation_is_email", "must be a valid email address")
Email = validation.NewStringRuleWithError(govalidator.IsEmail, ErrEmail)
```

Reply:
> How about adding `ErrorStruct() *Error` and drop the following two methods?

We can not use `ErrorStruct() *Error` because:
- If defining `func (r *MyRule) ErrorStruct() *Error`, this error affects all of the other uses of `MyRule`.
- If defining `func (r MyRule) ErrorStruct() *Error`, this error do not has any affect on our rule.

Reply:
Yeah, Predefining errors is very good, I change the source like what you said.